{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "r=requests.get('https://run.opencravat.org/submit/annotate?chrom=chr1&pos=2406483&ref_base=C&alt_base=G&annotators=aloft,cadd,cadd_exome,cancer_genome_interpreter,ccre_screen,chasmplus,civic,clingen,clinpred,clinvar,cosmic,cosmic_gene,cscape,dann,dann_coding,dbscsnv,dbsnp,dgi,ensembl_regulatory_build,ess_gene,exac_gene,fathmm,fathmm_xf_coding,funseq2,genehancer,gerp,ghis,gnomad,gnomad3,gnomad_gene,gtex,gwas_catalog,linsight,loftool,lrt,mavedb,metalr,metasvm,mutation_assessor,mutationtaster,mutpred1,mutpred_indel,ncbigene,ndex,ndex_chd,ndex_signor,omim,pangalodb,phastcons,phdsnpg,phi,phylop,polyphen2,prec,provean,repeat,revel,rvis,segway,sift,siphy,spliceai,uniprot,vest,cgc,cgd,varity_r&reports=vcf')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'revel': None,\n",
       " 'cadd_exome': None,\n",
       " 'cancer_genome_interpreter': None,\n",
       " 'ccre_screen': None,\n",
       " 'cgc': None,\n",
       " 'cgd': None,\n",
       " 'chasmplus': None,\n",
       " 'varity_r': None,\n",
       " 'clingen': None,\n",
       " 'clinpred': None,\n",
       " 'clinvar': {'sig': 'Likely pathogenic',\n",
       "  'disease_refs': 'MONDO:MONDO:0013936,MedGen:C3553947,OMIM:614870|MONDO:MONDO:0013937,MedGen:C3553948,OMIM:614871|MedGen:CN517202',\n",
       "  'disease_names': 'Peroxisome biogenesis disorder 6A (Zellweger)|Peroxisome biogenesis disorder 6B|not provided',\n",
       "  'rev_stat': 'criteria provided, single submitter',\n",
       "  'id': 554342,\n",
       "  'sig_conf': None},\n",
       " 'cosmic': None,\n",
       " 'cosmic_gene': {'gene_count': [['large intestine', '29'],\n",
       "   ['skin', '22'],\n",
       "   ['liver', '21'],\n",
       "   ['stomach', '13'],\n",
       "   ['lung', '12'],\n",
       "   ['haematopoietic and lymphoid tissue', '11'],\n",
       "   ['endometrium', '10'],\n",
       "   ['breast', '9'],\n",
       "   ['prostate', '9'],\n",
       "   ['central nervous system', '8'],\n",
       "   ['pancreas', '7'],\n",
       "   ['upper aerodigestive tract', '6'],\n",
       "   ['oesophagus', '5'],\n",
       "   ['kidney', '4'],\n",
       "   ['thyroid', '3'],\n",
       "   ['NS', '2'],\n",
       "   ['ovary', '2'],\n",
       "   ['urinary tract', '2'],\n",
       "   ['cervix', '1'],\n",
       "   ['soft tissue', '1'],\n",
       "   ['peritoneum', '1']],\n",
       "  'occurrences': 178},\n",
       " 'cscape': {'score': '0.978525'},\n",
       " 'dann': {'score': '0.73570852093793504\\r'},\n",
       " 'dann_coding': {'dann_coding_score': 0.9930931163127809,\n",
       "  'dann_rankscore': 0.58943},\n",
       " 'dbscsnv': {'ada_score': 0.99994568797423, 'rf_score': 0.926},\n",
       " 'dbsnp': {'rsid': 'rs1553231739'},\n",
       " 'dgi': None,\n",
       " 'ensembl_regulatory_build': {'region': 'promoter_flanking_region',\n",
       "  'ensr': 'ENSR00000918584'},\n",
       " 'ess_gene': None,\n",
       " 'exac_gene': {'exac_pli': 0.010463632475946,\n",
       "  'exac_prec': 0.945302377621403,\n",
       "  'exac_pnull': 0.0442339899026509,\n",
       "  'exac_nontcga_pli': 0.00759212357145324,\n",
       "  'exac_nontcga_prec': 0.925885534193925,\n",
       "  'exac_nontcga_pnull': 0.0665223422346215,\n",
       "  'exac_nonpsych_pli': 0.00748785882494236,\n",
       "  'exac_nonpsych_prec': 0.91248843970629,\n",
       "  'exac_nonpsych_pnull': 0.0800237014687672,\n",
       "  'exac_del_score': 0.229098528256408,\n",
       "  'exac_dup_score': -0.576593698772451,\n",
       "  'exac_cnv_score': -0.348380408944438,\n",
       "  'exac_cnv_flag': 'N'},\n",
       " 'fathmm': None,\n",
       " 'fathmm_xf_coding': None,\n",
       " 'funseq2': {'all': [['', '', '', '', '', '', '0.920173109385561']],\n",
       "  'hot': None,\n",
       "  'motif': None,\n",
       "  'score': '0.920173109385561'},\n",
       " 'genehancer': {'feature_name': 'Enhancer',\n",
       "  'score': 0.82,\n",
       "  'ident': 'GH01I002403',\n",
       "  'target_genes': 'PIR43593: 550.77, PEX10: 10.53, RER1: 7.11'},\n",
       " 'gerp': {'gerp_nr': 5.02, 'gerp_rs': 5.02, 'gerp_rs_rank': 0.66612},\n",
       " 'ghis': {'ghis': 0.546045359},\n",
       " 'gnomad': None,\n",
       " 'gnomad3': None,\n",
       " 'gnomad_gene': {'transcript': 'ENST00000507596',\n",
       "  'oe_lof': 0.4022,\n",
       "  'oe_mis': 1.0512,\n",
       "  'oe_syn': 0.95948,\n",
       "  'lof_z': 2.1519,\n",
       "  'mis_z': -0.26686,\n",
       "  'syn_z': 0.30038,\n",
       "  'pLI': 0.0075328,\n",
       "  'pRec': 0.9783,\n",
       "  'pNull': 0.014167,\n",
       "  'all': [['ENST00000288774',\n",
       "    0.51377,\n",
       "    1.0375,\n",
       "    0.97585,\n",
       "    1.7882,\n",
       "    -0.20114,\n",
       "    0.18559,\n",
       "    0.00034663,\n",
       "    0.95426,\n",
       "    0.04539],\n",
       "   ['ENST00000447513',\n",
       "    0.51655,\n",
       "    1.0478,\n",
       "    0.96747,\n",
       "    1.7731,\n",
       "    -0.24972,\n",
       "    0.24154,\n",
       "    0.00033523,\n",
       "    0.95228,\n",
       "    0.047383],\n",
       "   ['ENST00000507596',\n",
       "    0.4022,\n",
       "    1.0512,\n",
       "    0.95948,\n",
       "    2.1519,\n",
       "    -0.26686,\n",
       "    0.30038,\n",
       "    0.0075328,\n",
       "    0.9783,\n",
       "    0.014167],\n",
       "   ['ENST00000508384',\n",
       "    None,\n",
       "    1.0842,\n",
       "    0.83102,\n",
       "    None,\n",
       "    -0.1467,\n",
       "    0.45896,\n",
       "    None,\n",
       "    None,\n",
       "    None]]},\n",
       " 'gtex': None,\n",
       " 'gwas_catalog': None,\n",
       " 'linsight': {'value': 0.969215},\n",
       " 'loftool': {'loftool_score': 0.234},\n",
       " 'lrt': None,\n",
       " 'mavedb': None,\n",
       " 'metalr': None,\n",
       " 'metasvm': None,\n",
       " 'mutation_assessor': None,\n",
       " 'mutationtaster': {'transcript': 'ENST00000288774;ENST00000447513;ENST00000507596',\n",
       "  'score': 1.0,\n",
       "  'rankscore': 0.81001,\n",
       "  'prediction': 'Damaging',\n",
       "  'model': 'without_aae',\n",
       "  'all': [['ENST00000288774', 1.0, 0.81001, 'Damaging', 'without_aae'],\n",
       "   ['ENST00000447513', 1.0, 0.81001, 'Damaging', 'without_aae'],\n",
       "   ['ENST00000507596', 1.0, 0.81001, 'Damaging', 'without_aae']]},\n",
       " 'mutpred1': None,\n",
       " 'mutpred_indel': None,\n",
       " 'ncbigene': {'ncbi_desc': 'This gene encodes a protein involved in import of peroxisomal matrix proteins. This protein localizes to the peroxisomal membrane. Mutations in this gene result in phenotypes within the Zellweger spectrum of peroxisomal biogenesis disorders, ranging from neonatal adrenoleukodystrophy to Zellweger syndrome. Alternative splicing results in two transcript variants encoding different isoforms. [provided by RefSeq, Jul 2008]',\n",
       "  'entrez': '5192'},\n",
       " 'ndex': None,\n",
       " 'ndex_chd': None,\n",
       " 'ndex_signor': None,\n",
       " 'omim': {'omim_id': '614870; 614871'},\n",
       " 'pangalodb': None,\n",
       " 'phastcons': {'phastcons100_vert': 1.0,\n",
       "  'phastcons100_vert_r': 0.71638,\n",
       "  'phastcons30_mamm': 0.106,\n",
       "  'phastcons30_mamm_r': 0.19622,\n",
       "  'phastcons17way_primate': 0.255,\n",
       "  'phastcons17way_primate_r': 0.23391},\n",
       " 'phdsnpg': None,\n",
       " 'phi': {'phi': 0.10434},\n",
       " 'phylop': {'phylop100_vert': 7.397,\n",
       "  'phylop100_vert_r': 0.79162,\n",
       "  'phylop30_mamm': 0.931,\n",
       "  'phylop30_mamm_r': 0.33558,\n",
       "  'phylop17_primate': 0.599,\n",
       "  'phylop17_primate_r': 0.4025},\n",
       " 'polyphen2': None,\n",
       " 'prec': {'prec': 0.16529, 'stat': 'recessive'},\n",
       " 'provean': None,\n",
       " 'repeat': None,\n",
       " 'civic': None,\n",
       " 'rvis': {'rvis_evs': -0.27,\n",
       "  'rvis_perc_evs': 34.82,\n",
       "  'rvis_fdr_exac': 0.2265025,\n",
       "  'rvis_exac': 0.757984566,\n",
       "  'rvis_perc_exac': 86.30203},\n",
       " 'segway': {'sum_score': 14.438633, 'mean_score': 0.0103133092857},\n",
       " 'sift': None,\n",
       " 'siphy': {'pi': '0.0;1.0;0.0;0.0',\n",
       "  'logodds': 17.3042,\n",
       "  'logodds_rank': 0.87017},\n",
       " 'spliceai': {'ds_ag': 0.0004,\n",
       "  'ds_al': 0.0,\n",
       "  'ds_dg': 0.6946,\n",
       "  'ds_dl': 0.9962,\n",
       "  'dp_ag': 4,\n",
       "  'dp_al': 0,\n",
       "  'dp_dg': -4,\n",
       "  'dp_dl': 1},\n",
       " 'uniprot': {'acc': 'O60683'},\n",
       " 'vest': None,\n",
       " 'aloft': None,\n",
       " 'cadd': {'score': '5.251063', 'phred': '33'},\n",
       " 'crx': {'uid': '',\n",
       "  'chrom': 'chr1',\n",
       "  'pos': 2406483,\n",
       "  'ref_base': 'C',\n",
       "  'alt_base': 'G',\n",
       "  'note_variant': '',\n",
       "  'coding': '',\n",
       "  'hugo': 'PEX10',\n",
       "  'transcript': 'ENST00000447513.6',\n",
       "  'so': 'splice_site_variant',\n",
       "  'exonno': 6,\n",
       "  'cchange': 'c.912+1G>C',\n",
       "  'achange': '',\n",
       "  'all_mappings': '{\"PEX10\": [[\"O60683\", \"\", \"intron_variant,splice_site_variant\", \"ENST00000288774.7\", \"c.972+1G>C\"], [\"O60683\", \"\", \"intron_variant,splice_site_variant\", \"ENST00000447513.6\", \"c.912+1G>C\"], [\"\", \"\", \"intron_variant,splice_site_variant\", \"ENST00000507596.5\", \"c.912+1G>C\"], [\"\", \"\", \"2kb_downstream_variant,NMD_transcript_variant\", \"ENST00000510434.1\", \"c.*279G>C\"]], \"RER1\": [[\"\", \"\", \"2kb_downstream_variant\", \"ENST00000378513.7\", \"c.*3496C>G\"], [\"O15258\", \"\", \"2kb_downstream_variant\", \"ENST00000605895.6\", \"c.*3359C>G\"]]}',\n",
       "  'gposend': 2406483}}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "68"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(r.json().keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper function to flatten the nested dictionaries\n",
    "def flatten_dict(d, parent_key='', sep='.'):\n",
    "    items = {}\n",
    "    for key, value in d.items():\n",
    "        new_key = f'{parent_key}{sep}{key}' if parent_key else key\n",
    "        if isinstance(value, dict):\n",
    "            items.update(flatten_dict(value, new_key, sep=sep))\n",
    "        else:\n",
    "            items[new_key] = value\n",
    "    return items\n",
    "\n",
    "# Flatten the dictionary\n",
    "flat_data = flatten_dict(r.json())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>revel</th>\n",
       "      <th>cadd_exome</th>\n",
       "      <th>cancer_genome_interpreter</th>\n",
       "      <th>ccre_screen</th>\n",
       "      <th>cgc</th>\n",
       "      <th>cgd</th>\n",
       "      <th>chasmplus</th>\n",
       "      <th>varity_r</th>\n",
       "      <th>clingen</th>\n",
       "      <th>clinpred</th>\n",
       "      <th>...</th>\n",
       "      <th>crx.note_variant</th>\n",
       "      <th>crx.coding</th>\n",
       "      <th>crx.hugo</th>\n",
       "      <th>crx.transcript</th>\n",
       "      <th>crx.so</th>\n",
       "      <th>crx.exonno</th>\n",
       "      <th>crx.cchange</th>\n",
       "      <th>crx.achange</th>\n",
       "      <th>crx.all_mappings</th>\n",
       "      <th>crx.gposend</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PEX10</td>\n",
       "      <td>ENST00000447513.6</td>\n",
       "      <td>splice_site_variant</td>\n",
       "      <td>6</td>\n",
       "      <td>c.912+1G&gt;C</td>\n",
       "      <td></td>\n",
       "      <td>{\"PEX10\": [[\"O60683\", \"\", \"intron_variant,spli...</td>\n",
       "      <td>2406483</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 153 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  revel cadd_exome cancer_genome_interpreter ccre_screen   cgc   cgd  \\\n",
       "0  None       None                      None        None  None  None   \n",
       "\n",
       "  chasmplus varity_r clingen clinpred  ... crx.note_variant crx.coding  \\\n",
       "0      None     None    None     None  ...                               \n",
       "\n",
       "  crx.hugo     crx.transcript               crx.so crx.exonno crx.cchange  \\\n",
       "0    PEX10  ENST00000447513.6  splice_site_variant          6  c.912+1G>C   \n",
       "\n",
       "  crx.achange                                   crx.all_mappings crx.gposend  \n",
       "0              {\"PEX10\": [[\"O60683\", \"\", \"intron_variant,spli...     2406483  \n",
       "\n",
       "[1 rows x 153 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a DataFrame\n",
    "df = pd.DataFrame.from_dict(flat_data, orient='index').transpose()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENST</th>\n",
       "      <th>O60683.transcript</th>\n",
       "      <th>O60683.variant</th>\n",
       "      <th>.transcript</th>\n",
       "      <th>.variant</th>\n",
       "      <th>O15258.transcript</th>\n",
       "      <th>O15258.variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PEX10</td>\n",
       "      <td>ENST00000447513.6</td>\n",
       "      <td>c.912+1G&gt;C</td>\n",
       "      <td>ENST00000510434.1</td>\n",
       "      <td>c.*279G&gt;C</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RER1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENST00000378513.7</td>\n",
       "      <td>c.*3496C&gt;G</td>\n",
       "      <td>ENST00000605895.6</td>\n",
       "      <td>c.*3359C&gt;G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ENST  O60683.transcript O60683.variant        .transcript    .variant  \\\n",
       "0  PEX10  ENST00000447513.6     c.912+1G>C  ENST00000510434.1   c.*279G>C   \n",
       "1   RER1               None           None  ENST00000378513.7  c.*3496C>G   \n",
       "\n",
       "   O15258.transcript O15258.variant  \n",
       "0               None           None  \n",
       "1  ENST00000605895.6     c.*3359C>G  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract data from 'crx' dictionary\n",
    "crx_data = r.json().get('crx', {})\n",
    "\n",
    "# Parse 'all_mappings' JSON and extract ENST mappings\n",
    "all_mappings_json = crx_data.get('all_mappings', '{}')\n",
    "all_mappings = json.loads(all_mappings_json)\n",
    "\n",
    "# Create a list of dictionaries where each dictionary represents an ENST mapping\n",
    "enst_data = []\n",
    "for enst, mappings in all_mappings.items():\n",
    "    enst_dict = {'ENST': enst}\n",
    "    for mapping in mappings:\n",
    "        gene, _, _, transcript, variant = mapping\n",
    "        enst_dict[f'{gene}.transcript'] = transcript\n",
    "        enst_dict[f'{gene}.variant'] = variant\n",
    "    enst_data.append(enst_dict)\n",
    "\n",
    "# Create a DataFrame from the list of ENST mappings\n",
    "df = pd.DataFrame(enst_data)\n",
    "\n",
    "# Fill NaN values with None\n",
    "df = df.fillna('None')\n",
    "df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "training",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
